cmd/watch.go

```diff
@@ -0,0 +1,85 @@
+package cmd
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+
+	"github.com/redhat-developer/odo/pkg/application"
+	"github.com/redhat-developer/odo/pkg/component"
+	"github.com/redhat-developer/odo/pkg/project"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var watchLocal string
+
+var watchCmd = &cobra.Command{
+	Use:   "watch [component name]",
+	Short: "Watch for changes, update component on change",
+	Long:  `Watch for changes, update component on change.`,
+	Example: `  # Watch for changes in directory for current component
+  ocdev watch
+
+  # Watch for changes in directory for component called frontend 
+  ocdev watch frontend
+	`,
+	Args: cobra.MaximumNArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		stdout := os.Stdout
+		client := getOcClient()
+		projectName := project.GetCurrent(client)
+		applicationName, err := application.GetCurrent(client)
+		checkError(err, "Unable to get current application.")
+
+		var componentName string
+		if len(args) == 0 {
+			var err error
+			log.Debug("No component name passed, assuming current component")
+			componentName, err = component.GetCurrent(client)
+			if err != nil {
+				checkError(err, "")
+			}
+			if componentName == "" {
+				fmt.Println("No component is set as active.")
+				fmt.Println("Use 'ocdev component set <component name> to set and existing component as active or call this command with component name as and argument.")
+				os.Exit(1)
+			}
+		} else {
+			componentName = args[0]
+		}
+
+		sourceType, sourcePath, err := component.GetComponentSource(client, componentName, applicationName, projectName)
+		checkError(err, "Unable to get source for %s component.", componentName)
+		if sourceType != "local" {
+			fmt.Println("You use watch only with components with local source.")
+			os.Exit(1)
+		}
+
+		var watchPath string
+		if watchLocal != "" {
+			watchPath = watchLocal
+		} else {
+			u, err := url.Parse(sourcePath)
+			checkError(err, fmt.Sprintf("Unable to parse source %s from component %s.", sourcePath, componentName))
+
+			if u.Scheme != "" && u.Scheme != "file" {
+				fmt.Printf("Component %s has invalid source path %s.", componentName, u.Scheme)
+				os.Exit(1)
+			}
+			watchPath = u.Path
+		}
+
+		err = component.WatchAndPush(client, componentName, watchPath, stdout)
+		if err != nil {
+			checkError(err, "Error while trying to watch %s", watchPath)
+		}
+	},
+}
+
+func init() {
+	watchCmd.Flags().StringVar(&watchLocal, "local", "", "Use given local directory as a source for component. (It must be a local component)")
```

Comment:
@kadel Should we add another parameter for setting the delay?

Reply:
I don't want to complicate it right from the beginning. If there is a need for it we can add it later.

Reply:
What is the use-case for waiting before syncing?  Any downside to removing the delay? Is there any way to remove the delay entirely? 

As a developer, I use "watch" to establish a fast inner dev loop.  Slower tooling is much less useful to me.

Reply:
Hi @ryanj 

> What is the use-case for waiting before syncing? Any downside to removing the delay? 

In case of some intensive or batch operation (like a building) there might be a lot of changes, if we would trigger sync immediately after each change it would be ineffective. 

`oc rsync` also uses delayed sync (3 seconds). To be honest I don't remember why I increased it to 5s  😬 

> Is there any way (as a user) to remove the delay entirely?

Not right now. But if there is a real use case for this we could remove delay, and make it configurable via commandline argument

This is just the first implementation mainly to collect feedback. 

> oc rsync already watches for changes. Is another watcher needed?

In future we would like to remove any dependencies on `oc`  binary. Right now it uses `oc rsync` for syncing operation because it was just the quickest way to implement this.